controllers/ndbserver_controller.go

```diff
@@ -37,26 +45,90 @@ type NDBServerReconciler struct {
 //+kubebuilder:rbac:groups=ndb.nutanix.com,resources=ndbservers/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=ndb.nutanix.com,resources=ndbservers/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the NDBServer object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+/*
+Reconciles the NDBServer custom resources by
+1. Checks for deletion
+2. Verify credentials and connectivity
+3. Take actions based on current status.status, fetch data
+4. Update the status if any changes are observed (excluding counter)
+*/
 func (r *NDBServerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := log.FromContext(ctx)
+	log.Info("NDBServer reconcile started")
+	// 1. Checks for deletion
+	// Fetch the NDBServer resource from the namespace
+	ndbServer := &ndbv1alpha1.NDBServer{}
+	err := r.Get(ctx, req.NamespacedName, ndbServer)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Return and don't requeue
+			log.Info("NDBServer resource not found. Ignoring since object must be deleted")
+			return doNotRequeue()
+		}
+		// Error reading the object - requeue the request.
+		log.Error(err, "Failed to get NDBServer")
+		return requeueOnErr(err)
+	}
+
+	// Copy the status, this will be updated by the end of the reconcile
+	status := ndbServer.Status.DeepCopy()
+	// Initialize maps if they do not exist
+	if status.Databases == nil {
+		status.Databases = make(map[string]ndbv1alpha1.NDBServerDatabaseInfo)
+	}
+
+	// log.Info("NDBServer CR Status: " + util.ToString(status))
+
+	// 2. Verify credentials and connectivity
+	// Fetch credentials and check Authentication
+	var ndbClient *ndb_client.NDBClient
+	username, password, caCert, err := getNDBCredentialsFromSecret(ctx, r.Client, ndbServer.Spec.CredentialSecret, req.Namespace)
+	if err != nil {
+		log.Error(err, "Credential Error: error while fetching credentials from CredentialSecret", "secret name", ndbServer.Spec.CredentialSecret)
+		status.Status = common.NDB_CR_STATUS_CREDENTIAL_ERROR
+	} else {
+		ndbClient = ndb_client.NewNDBClient(username, password, ndbServer.Spec.Server, caCert, ndbServer.Spec.SkipCertificateVerification)
+		authResponse, err := ndb_api.AuthValidate(ctx, ndbClient)
+		if err != nil || authResponse.Status != common.AUTH_RESPONSE_STATUS_SUCCESS {
```

Comment:
possible to check status code eg 200, instead of string compare?

Reply:
I see what you're trying to say here, checking 200 would be a better approach too. But in the current implementation of ndb_api package methods the status code is not returned. We can make changes to that package to add a return value of the status too and make necessary re-factorings. I'm thinking of taking that up as a separate task/PR to not increase the scope of this one, WDYT?

Reply:
Sure, just leave a TODO comment on top of that method.